### PR TITLE
loopEither - a more generic method of loop

### PR DESCRIPTION
**Pull request by nikiforo**
I don't have a strong opinion whether this function should exist in FS2. Thus, I submit this to hear different views/opinions.

Personally I wrote it for error handling, but I think that my case can be implemented with https://github.com/cb372/cats-retry.

### CODE REVIEW

File: core/shared/src/test/scala/fs2/PullSuite.scala [hunk 1]
```diff
@@ -50,4 +50,39 @@ class PullSuite extends Fs2Suite {
       }
     }
   }
+
+  property("loop") {
+    forAll { (list: List[Int]) =>
+      def customOutput(s: List[Int]): Pull[Fallible, Int, Option[List[Int]]] =
```
Comment (nikiforo) [inline core/shared/src/test/scala/fs2/PullSuite.scala:56]: `Pure` is enough here

File: core/shared/src/test/scala/fs2/PullSuite.scala [hunk 2]
```diff
@@ -50,4 +50,39 @@ class PullSuite extends Fs2Suite {
...
+
+      val result = Pull.loop(customOutput)(list).stream.compile.toList
+
+      assertEquals(result, Right(list))
+    }
+  }
+
+  property("loopEither") {
+    forAll { (list: List[Int]) =>
+      def customOutputReturnsString(s: List[Int]): Pull[Fallible, Int, Either[List[Int], String]] =
```
Comment (nikiforo) [inline core/shared/src/test/scala/fs2/PullSuite.scala:70]: `Pure` is enough here

File: core/shared/src/main/scala/fs2/Pull.scala [hunk 3]
```diff
@@ -414,11 +414,16 @@ object Pull extends PullLowPriority {
...
     */
-  def loop[F[_], O, R](f: R => Pull[F, O, Option[R]]): R => Pull[F, O, Unit] =
-    (r: R) =>
-      f(r).flatMap {
-        case None    => Pull.done
-        case Some(s) => loop(f)(s)
+  def loop[F[_], O, R](f: R => Pull[F, O, Option[R]]): R => Pull[F, O, Unit] = {
+    val ef = (r: R) => f(r).map(_.fold(().asRight[R])(_.asLeft))
+    loopEither(ef)
+  }
```
Comment (diesalbla) [inline core/shared/src/main/scala/fs2/Pull.scala:420]: I would prefer to keep `loop` as defined on its own.

**Comment by mpilquist** [comment]
Looks good to me - let's keep original definition of `loop` like suggested by @diesalbla and add ScalaDoc.

**Pull request merged**